File: GenerateParentheses/GenerateParentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        subgenerate(result, "", n, 0);
        return result;
    }

private:
	void subgenerate(vector<string> &result, string subStr, int left, int right) {
		if(left == 0 && right == 0) {
			result.push_back(subStr);
			return;
		}
		if(left > 0)
			subgenerate(result, subStr + '(', left - 1, right + 1);
		if(right > 0)
			subgenerate(result, subStr + ')', left, right - 1);
	}
};

File: GenerateParentheses/GenerateParentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = []
        generate("", n, 0, res)
        return res

def generate(st, left, right, res):
    if left == 0 and right == 0:
        return res.append(st)
    if left > 0:
        generate(st+'(', left-1, right+1, res)
    if right > 0:
        generate(st+')', left, right-1, res)